### PR TITLE
suppressions found thus far

### LINKED ISSUE #16202: Update Valgrind suppressions for meowbuild
## Reason
New suppressions needed when performing Valgrind on meowbuild machines

## Design
Run through [tutorial](https://wiki.wxwidgets.org/Valgrind_Suppression_File_Howto) to build suppressions. 

## Impact
Moving forwards to testing on new build farm.

### PR DESCRIPTION
**Pull request by milljm**
Adding specific system calls causing memory leaks while operating on meowbuild.

Closes #16202

### CODE REVIEW

**Comment by loganharbour** [comment]
@milljm what target is this for?

**Comment by milljm** [comment]
oprof? if that was your question, fixing these errors:

https://civet.inl.gov/job/620371/

File: python/TestHarness/suppressions/errors.supp [hunk 1]
```diff
@@ -189,3 +189,50 @@
    fun:*boost*math*
    ...
 }
+{
+   meowbuild_noai6ai_cached
```
Comment (permcody) [inline python/TestHarness/suppressions/errors.supp:193]: I don't think you should name these new suppressions "meowbuild". Can you name them after the OS or some other more useful trait of the system?

**Review by permcody — CHANGES_REQUESTED** [review]
I don't see any new suppressions that concern me. As long as they are not in our stack it should be fine to ignore them. I think the names should be changed though.

File: python/TestHarness/suppressions/errors.supp (hunk 1 above)
Comment (milljm) [reply python/TestHarness/suppressions/errors.supp:193]: spack perhaps? All the files being scrutinized are originating from the the compiler stack supplied by a spack path.

**Comment by lindsayad** [comment]
I don't think we should be adding suppressions for reachable memory. See [here](https://stackoverflow.com/a/6973510/4493669). The default leak kinds that are valgrind errors are `definite` and `possible`, and we do not currently modify that in our valgrind testing (the valgrind option for that would be `--show-leak-kinds=<definite,possible,indirect,reachable,all>`. I think we should just stick with the defaults, which would allows us to erase almost all the suppressions added here.

**Comment by lindsayad** [comment]
For the invalid free involving `noi6ai_cached`, I'm seeing:

- https://github.com/sparklemotion/nokogiri/issues/1654#issuecomment-309429215
- https://bugs.gentoo.org/472544
- https://svn.boost.org/trac10/ticket/12188

So this definitely seems like a good suppression

**Comment by lindsayad** [comment]
Appears to be a glibc bug: https://bugzilla.redhat.com/show_bug.cgi?id=754026

**Comment by lindsayad** [comment]
@milljm I'm sorry if I should know this answer, but is there a reason we are using what appears to be version 2.17 of glibc on meowbuild? That version of glibc was released in 2012.

**Comment by lindsayad** [comment]
> @milljm I'm sorry if I should know this answer, but is there a reason we are using what appears to be version 2.17 of glibc on meowbuild? That version of glibc was released in 2012.

I'm guessing it's a spack thing...? In that case, then for sure suppression all the way! (As opposed to upgrading glibc version)

**Comment by milljm** [comment]
> > @milljm I'm sorry if I should know this answer, but is there a reason we are using what appears to be version 2.17 of glibc on meowbuild? That version of glibc was released in 2012.
> 
> I'm guessing it's a spack thing...? In that case, then for sure suppression all the way! (As opposed to upgrading glibc version)

```pre
==2253143== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
--2253142-- Discarding syms at 0x17b798f0-0x17b88072 in /usr/lib64/libresolv-2.17.so (have_dinfo 1)
==2253142== Invalid free() / delete / delete[] / realloc()
==2253142==    at 0x4C2AF9D: free (vg_replace_malloc.c:540)
==2253142==    by 0x111CC8BB: __libc_freeres (in /usr/lib64/libc-2.17.so)
==2253142==    by 0x4A24739: _vgnU_freeres (vg_preloaded.c:77)
==2253142==    by 0x11095C5A: __run_exit_handlers (in /usr/lib64/libc-2.17.so)
==2253142==    by 0x11095CE6: exit (in /usr/lib64/libc-2.17.so)
==2253142==    by 0x1107E50B: (below main) (in /usr/lib64/libc-2.17.so)
==2253142==  Address 0x114233c0 is 0 bytes inside data symbol "noai6ai_cached"
==2253142== 
{
   <insert_a_suppression_name_here>
   Memcheck:Free
   fun:free
   fun:__libc_freeres
   fun:_vgnU_freeres
   fun:__run_exit_handlers
   fun:exit
   fun:(below main)
}
```

Looks more like a CentOS thing in this case. If it were Spack, I would have much more control over what version to use.

**Comment by milljm** [comment]
It might be best to upgrade from CentOS 7 to 8. There's a task...

**Comment by lindsayad** [comment]
> It might be best to upgrade from CentOS 7 to 8. There's a task...

A task that can wait for another time. Suppressing this is a good solution in the meantime

**Comment by milljm** [comment]
Well, I've already started. And I knew we would need to support this feature in Civet with the new machines. A day or two, and think I can try the Valgrind tests on this new container (CentOS 8).

**Comment by milljm** [comment]
By switching to CentOS 8, I've reduced the amount of suppressions I need (two). But a new one showed up, which does look like something of ours:

```pre
==1851306== LEAK SUMMARY:
==1851306==    definitely lost: 0 bytes in 0 blocks
==1851306==    indirectly lost: 0 bytes in 0 blocks
==1851306==      possibly lost: 0 bytes in 0 blocks
==1851306==    still reachable: 0 bytes in 0 blocks
==1851306==         suppressed: 779,139 bytes in 8,928 blocks
==1851306== 
==1851306== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
--1851306-- 
--1851306-- used_suppression:     12 meowbuild_mpich_libmpi /scratch/moose/python/TestHarness/suppressions/errors.supp:201 suppressed: 656,043 bytes in 8,920 blocks
--1851306-- used_suppression:      8 meowbuild_noai6ai_cached /scratch/moose/python/TestHarness/suppressions/errors.supp:193 suppressed: 123,096 bytes in 8 blocks
==1851306== 
==1851306== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
==1851305== Checked 4,936,304 bytes
==1851305== 
==1851305== 5 bytes in 1 blocks are still reachable in loss record 2 of 22
==1851305==    at 0x4C4404A: malloc (vg_replace_malloc.c:309)
==1851305==    by 0x8E7D717: NCDISPATCH_initialize (in /scratch/moose/libmesh/installed/lib/libnetcdf.so.13.1.1)
==1851305==    by 0x8E71ED6: nc_initialize (in /scratch/moose/libmesh/installed/lib/libnetcdf.so.13.1.1)
==1851305==    by 0x8E75AC0: NC_open (in /scratch/moose/libmesh/installed/lib/libnetcdf.so.13.1.1)
==1851305==    by 0x8E74EA3: nc_open (in /scratch/moose/libmesh/installed/lib/libnetcdf.so.13.1.1)
==1851305==    by 0x86F346D: ex_open_int (ex_open.c:165)
==1851305==    by 0x8157E3B: libMesh::ExodusII_IO_Helper::open(char const*, bool) (exodusII_io_helper.C:516)
==1851305==    by 0x814408B: libMesh::ExodusII_IO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (exodusII_io.C:156)
==1851305==    by 0x82EBC70: libMesh::NameBasedIO::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (namebased_io.C:219)
==1851305==    by 0x83381C1: libMesh::UnstructuredMesh::read(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, void*, bool, bool) (unstructured_mesh.C:637)
==1851305==    by 0x651B47D: FileMesh::buildMesh() (FileMesh.C:119)
==1851305==    by 0x6513200: MooseMesh::init() (MooseMesh.C:2369)
==1851305== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   match-leak-kinds: reachable
   fun:malloc
   fun:NCDISPATCH_initialize
   fun:nc_initialize
   fun:NC_open
   fun:nc_open
   fun:ex_open_int
   fun:_ZN7libMesh18ExodusII_IO_Helper4openEPKcb
   fun:_ZN7libMesh11ExodusII_IO4readERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
   fun:_ZN7libMesh11NameBasedIO4readERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
   fun:_ZN7libMesh16UnstructuredMesh4readERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPvbb
   fun:_ZN8FileMesh9buildMeshEv
   fun:_ZN9MooseMesh4initEv
}
```

**Comment by lindsayad** [comment]
> ==1851306== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
> --1851306-- 
> --1851306-- used_suppression:     12 meowbuild_mpich_libmpi /scratch/moose/python/TestHarness/suppressions/errors.supp:201 suppressed: 656,043 bytes in 8,920 blocks
> --1851306-- used_suppression:      8 meowbuild_noai6ai_cached /scratch/moose/python/TestHarness/suppressions/errors.supp:193 suppressed: 123,096 bytes in 8 blocks
> ==1851306== 
> ==1851306== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
> ==1851305== Checked 4,936,304 bytes
> ==1851305== 
> ==1851305== 5 bytes in 1 blocks are still reachable in loss record 2 of 22

We don't consider `still reachable` to be an error, hence why valgrind is telling us we have `0 errors from 0 contexts`. Does MOOSE register the test as failed without that `still reachable` suppression? If it does, that would be unexpected

**Comment by milljm** [comment]
Okay... awesome. By switching to CentOS 8, all the memory leaks went away.

Cheers!

**Comment by lindsayad** [comment]
Awesome. Awesome that you got it done so fast!

**Pull request closed without merging**